Hart Auto-Load

yezr opened the pull request:
fixes #918. Cribs a lot of the logic from #908

yezr commented:
The logic works, but I would love a review of the code syntax.

artoonie commented on src/main/java/network/brightspots/rcv/HartCvrReader.java:
```diff
@@ -109,10 +117,31 @@ private void readCastVoteRecord(List<CastVoteRecord> castVoteRecords, Path path)
+              candidateCode = Tabulator.UNDECLARED_WRITE_IN_OUTPUT_LABEL;
+            }
+            if (this.candidateCodesToCandidates.containsKey(candidateCode)) {
+              if (!this.candidateCodesToCandidates.get(candidateCode).Name.equals(candidateName)) {
```
These two ifs can be simplified a bit:

```
Candidate existingCandidate = this.candidateCodesToCandidates.get(candidateCode);
if (existingCandidate != null && existingCandidate.Name.equals(candidateName)) {
```

yezr replied:
along with the computeIfAbsent update this should be addressed

artoonie commented on src/main/java/network/brightspots/rcv/HartCvrReader.java:
```diff
@@ -109,10 +117,31 @@ private void readCastVoteRecord(List<CastVoteRecord> castVoteRecords, Path path)
+                        Candidate Code %s associated with more than one candidate name.
+                        Originally associated with name '%s'.
+                        In CVR at '%s' it is associated with '%s'.
+                        """.formatted(candidateCode,
```
This syntax will cause the error message to have a ton of spaces throughout the error message. Using `+`s will avoid that:
```
                String message =
                        "Candidate Code %s associated with more than one candidate name.  " +
                        "Originally associated with name '%s'. " + 
                        "In CVR at '%s' it is associated with '%s'."
                        .formatted(candidateCode,
```

artoonie commented on src/main/java/network/brightspots/rcv/HartCvrReader.java:
```diff
@@ -109,10 +117,31 @@ private void readCastVoteRecord(List<CastVoteRecord> castVoteRecords, Path path)
+                throw new CastVoteRecord.CvrParseException();
+              }
+            } else {
+              if (!candidateCode.equals(Tabulator.UNDECLARED_WRITE_IN_OUTPUT_LABEL)) {
```
Combine into `else if {` rather than `else { if {`

artoonie commented on src/main/java/network/brightspots/rcv/HartCvrReader.java:
```diff
@@ -109,10 +117,31 @@ private void readCastVoteRecord(List<CastVoteRecord> castVoteRecords, Path path)
+              if (!candidateCode.equals(Tabulator.UNDECLARED_WRITE_IN_OUTPUT_LABEL)) {
+                this.candidateCodesToCandidates.put(candidateCode,
+                        new Candidate(candidateName, candidateCode));
+              }
```
I think you can simplify this whole block using `computeIfAbsent`:

```
          for (Option option : contest.Options) {
            final String candidateCode = option.Id;
            final String candidateName = option.Name;
            final boolean isUndeclaredWriteIn = candidateCode.equals(
                    source.getUndeclaredWriteInLabel());

            if (!isUndeclaredWriteIn) {
              this.candidateCodesToCandidates.computeIfAbsent(candidateCode, k ->
                      new Candidate(candidateName, candidateCode));

              String candidateNameFromMap = this.candidateCodesToCandidates.get(candidateCode).Name;
              if (!candidateNameFromMap.equals(candidateName)) {
                String message = String.format(
                        "Candidate Code %s associated with more than one candidate name. "
                                + "Originally associated with name '%s'. "
                                + "In CVR at '%s' it is associated with '%s'.",
                        candidateCode, candidateNameFromMap, path.getFileName(), candidateName
                );
                Logger.severe(message);
                throw new CastVoteRecord.CvrParseException();
              }
            }

            // Hart RCV election ranks are indicated by a string read left to right:
            // each digit corresponds to a rank and is set to 1 if that rank was voted:
            // 0100 indicates rank 2 was voted
            // 0000 indicates no rank was voted (undervote)
            // 0101 indicates ranks 2 and 4 are voted (overvote)
            for (int rank = 1; rank < option.Value.length() + 1; rank++) {
              String rankValue = option.Value.substring(rank - 1, rank);
              if (rankValue.equals("1")) {
                if (isUndeclaredWriteIn) {
                  rankings.add(new Pair<>(rank, UNDECLARED_WRITE_IN_OUTPUT_LABEL));
                } else {
                  rankings.add(new Pair<>(rank, candidateCode));
                }
              }
            }
          }
```

yezr replied:
great heads up. I updated to use this to put the candidate code/name into the lookup if the code isn't there. then use the lookup to compare against the current code/name to see if the current CVR's candidate code/name matches the lookup. if not, it throws.

artoonie replied:
This looks good to me

artoonie commented on src/main/java/network/brightspots/rcv/HartCvrReader.java:
```diff
@@ -147,6 +176,31 @@ private void readCastVoteRecord(List<CastVoteRecord> castVoteRecords, Path path)
+      } catch (CastVoteRecord.CvrParseException e) {
+        Logger.severe("Error gathering Unknown Candidates\n%s", e);
+        return new HashSet<>();
+      } catch (IOException e) {
```
You can collapse the two catches into one:
```
      } catch (CastVoteRecord.CvrParseException | IOException e) {
```

artoonie commented on src/main/java/network/brightspots/rcv/HartCvrReader.java:
```diff
@@ -147,6 +176,31 @@ private void readCastVoteRecord(List<CastVoteRecord> castVoteRecords, Path path)
+    Set<String> knownNames = config.getCandidateNames();
+    if (this.candidateCodesToCandidates.entrySet().isEmpty()) {
+      try {
+        //Reading the CVRs will load this.candidateCodestoCandidates
```
```suggestion
        // Reading the CVRs will load this.candidateCodestoCandidates
```

yezr commented:
@artoonie I addressed all your comments. I closed all the ones that I don't think need further discussion. There are two open that implement the map.ComputeIfAbsent you suggested that I'm interested in your feedback on

artoonie commented on src/main/java/network/brightspots/rcv/GuiConfigController.java:
```diff
@@ -1905,8 +1905,13 @@ protected LoadedCvrData call() {
+              if (!isNullOrBlank(e.getMessage())) {
+                message += ": " + e.getMessage();
+              }
+              Logger.severe(message);
```
If getting an NPE here, I believe this could be simplified with:
`Logger.severe("Failed to read CVRs: " + e.getMessage());`
which would just say "Failed to read CVRs: null" if null. (And that's easier to include in the list of errors that we need for the TDP)

yezr replied:
I was thinking that any place where we show `null` explicitly to an end user could be alarming and I was trying to avoid that. 

I didn't even think about the list of errors reporting requirements. But I think here either way we are going to have to search down the different `e.getMessage()` potential descriptions. The only difference with this change would be that the case where there is no explicit message is a little bit different. I think it's worth it to remove the `null` reference.

artoonie commented on src/main/java/network/brightspots/rcv/HartCvrReader.java:
```diff
@@ -109,10 +118,31 @@ private void readCastVoteRecord(List<CastVoteRecord> castVoteRecords, Path path)
-              candidateName = Tabulator.UNDECLARED_WRITE_IN_OUTPUT_LABEL;
+
+            //Can be null with some write-ins
+            String candidateNameClean = Objects.requireNonNullElse(option.Name, "");
```
How far can we get without setting this to empty string? If it's null, I think it makes more sense to keep it null as long as possible -- unless that causes a ton of issues everywhere. Another idea is to set the null candidate name to the ID instead.

artoonie commented on src/main/java/network/brightspots/rcv/HartCvrReader.java:
```diff
@@ -109,10 +118,31 @@ private void readCastVoteRecord(List<CastVoteRecord> castVoteRecords, Path path)
+              candidate.Code = Tabulator.UNDECLARED_WRITE_IN_OUTPUT_LABEL;
+            } else {
+              this.candidateCodesToCandidates.putIfAbsent(candidate.Code,
+                      candidate);
```
[nit] I think these two lines would fit on one line

artoonie commented on src/main/java/network/brightspots/rcv/HartCvrReader.java:
```diff
@@ -147,6 +177,28 @@ private void readCastVoteRecord(List<CastVoteRecord> castVoteRecords, Path path)
+            .map(entry -> new RawContestConfig.Candidate(entry.getValue().Name, entry.getKey()))
+            .collect(Collectors.toSet());
+  }
+
```
This code seems useful to move to `BaseCvrReader` to replace `gatherUnknownCandidates` there -- any downsides to that I don't see?

yezr replied:
Started digging in to this. When trying to move it to `BaseCvrReader` it looks like it would need to know about the `candidateCodesToCandidates` map. The reason is because that map is lazy loaded. If we get here and it is empty we take a pass through the CVRs to populate it. Once it's populated we use it for the comparison to see which haven't been configured.

artoonie replied:
Looked into this and agree -- not worth changing. The two Candidate classes are different, so any consolidation would be a bigger lift. Keep as is!
